### ISSUE TITLE
forall-ensures not quite equivalent to assert-by

### ISSUE THREAD

**Issue by davidcok** [dafny-lang/dafny#3559]
### Dafny version

3.11.1+dev

### Code to produce this issue

```dafny
The NonlinearArithmetic/Internals/ModInternals.dfy file in the library
```

### Command to run and resulting output

```code
There are 4 instances of forall ensures.
With dafny verify --function-syntax:4 it verifies fine; but with forall ensures replaced by assert-by, as recommended in the warning that is emitted, the file no longer verifies.
```

### What happened?

If the two syntaxes are equivalent they need to behave the same way for proofs.

### What type of operating system are you experiencing the problem on?

Mac

**Comment by RustanLeino** [comment]
I have looked at these encodings and played with the example. My conclusion is that the problem with `ModInternals.dfy` arises from minute differences in what gets sent to the SMT solver, which affects the already brittle nature of dealing with multiplication. When I have tried to isolate these problems into a separate, smaller file, then everything verifies.

So, I'm still happy with deprecating the degenerate form of the `forall` statement. Until someone writes an alternative proof for the four lemmas in `ModInternals.dfy`, we'll be getting the deprecation warning on the degenerate `forall` statements that those lemmas currently use.

**Comment by MikaelMayer** [comment]
Indeed, if we generate the boogie files for the two cases, one for assert by and the other for forall ensures .... both verify. So it's actually one more discrepancy between the boogie file that is generated/parsed and what Dafny sends to Boogie.
